src/redux/books/booksSlice.js

```diff
@@ -1,45 +1,136 @@
-  value: [
-    {
-      item_id: 'item1',
-      title: 'The Great Gatsby',
-      author: 'John Smith',
-      category: 'Fiction',
-    },
-    {
-      item_id: 'item2',
-      title: 'Anna Karenina',
-      author: 'Leo Tolstoy',
-      category: 'Fiction',
-    },
-    {
-      item_id: 'item3',
-      title: 'The Selfish Gene',
-      author: 'Richard Dawkins',
-      category: 'Nonfiction',
-    },
-  ],
+  books: [],
+  loading: false,
+  error: null,
 };
 
+// Get
+export const fetchBooks = createAsyncThunk(
+  'books/fetchBooks',
+  async (_, { rejectWithValue }) => {
+    try {
+      const response = await axios.get(`${baseUrl}${appId}/books`);
+      return response.data;
+    } catch (error) {
+      return rejectWithValue(error.message);
+    }
+  },
+);
+
+// Post
+export const addBook = createAsyncThunk(
+  'books/addBook',
+  async (bookData, { rejectWithValue }) => {
+    try {
+      const response = await axios.post(`${baseUrl}${appId}/books`, {
+        item_id: bookData.item_id,
+        title: bookData.title,
+        author: bookData.author,
+        category: bookData.category,
+      });
+      return response.data;
+    } catch (error) {
+      return rejectWithValue(error.message);
+    }
+  },
+);
+
+// Delete
+export const removeBook = createAsyncThunk(
+  'books/removeBook',
+  async (payload, { rejectWithValue }) => {
+    try {
+      await axios.delete(`${baseUrl}${appId}/books/${payload.itemId}`, {
+        data: {
+          item_id: payload.itemId,
+        },
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      });
+      return { itemId: payload.itemId };
+    } catch (error) {
+      return rejectWithValue(error.message);
+    }
+  },
+);
+
 const booksSlice = createSlice({
   name: 'books',
   initialState,
-  reducers: {
-    add: (state, action) => {
-      const { title, author } = action.payload;
-      state.value.push({
-        item_id: `item${Math.floor(Math.random() * 10000)}`,
-        title,
-        author,
+  reducers: {},
+  extraReducers: (builder) => {
+    builder
+      .addCase(fetchBooks.pending, (state) => {
+        state.loading = true;
+        state.error = null;
+      })
+      .addCase(fetchBooks.fulfilled, (state, action) => {
+        state.loading = false;
+        state.books = action.payload;
+      })
+      .addCase(fetchBooks.rejected, (state, action) => {
+        state.loading = false;
+        state.error = action.payload;
+      })
+      .addCase(addBook.pending, (state) => {
+        state.loading = true;
+        state.error = null;
+      })
+      .addCase(addBook.fulfilled, (state, action) => {
+        state.loading = false;
+        const newItem = action.payload;
+        state.books[newItem.item_id] = [
+          {
+            item_id: newItem.item_id,
+            title: newItem.title,
+            author: newItem.author,
+            category: newItem.category,
+          },
+        ];
+      })
```

Comment:
As suggested by the previous reviewer kindly fix the `addBook` function so that it is possible to add books upon form submission. Currently I'm getting the following error while trying to add the books.

<img width="1242" alt="Screenshot 2023-07-01 at 4 37 07 PM" src="https://github.com/M-AminAlizadeh/microverse-book-store/assets/44798044/b0b0879b-db1b-44f8-9f89-cc86b056a225">


In order to fix the problem you can fix the returned value of your addBook function as following:

```suggestion
      .addCase(addBook.fulfilled, (state, action) => {
        state.loading = false;
       return[... {
            item_id: action.payload.item_id,
            title: action.payload.title,
            author: action.payload.author,
            category: action.payload.category,
          },
        ];
      })
```